bpf/bpf_lxc.c

```diff
@@ -850,6 +850,27 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 skip_egress_gateway:
 #endif
 
+#ifdef ENABLE_VTEP
+	{
+		int i;
+		mac_t vtep_mac = 0;
+
+		for (i = 0; i < VTEP_NUMS; i++) {
+			if (tunnel_endpoint == VTEP_ENDPOINT[i]) {
+				vtep_mac = VTEP_MAC[i];
+				break;
+			}
+		}
```

Comment:
Couldn't these VTEP endpoints and MAC addresses be stored in a dedicated hashmap (as keys and values respectively)? That would make the lookup O(1) instead of O(n) and would avoid the limit on `VTEP_NUMS`.

Reply:
for now  I am limiting the number of VTEPs to 2 due to https://github.com/cilium/ebpf/issues/553, and it is common like in BIG-IP case to have active/standby VTEP device as pair, it is possible in the future users could asking for supporting more VTEPs devices, I am thinking using `_Pragma("unroll")` for the for loop. I don't know much about hashmap, and how to define the hashmap in macro in `node_config.h`, that seems bit of complicate :), but yes, i will  experiment with it as follow-up

Reply:
You don't need to implement the hashmap yourself. It's already a type of BPF map. Grep for `BPF_MAP_TYPE_HASH` in our BPF source code. Most of our BPF maps are hashmaps. You then wouldn't need to limit the number of VTEPs and to unroll the loop.

Reply:
I see what you mean, yes i think Joe at one point has shared similar idea, the initial implementation of this PR is to use tunnel_map, but it seems tunnel_map will be deprecated, and the suggestion is to use ipcache map, I have thought about creating a BPF map like tunnel_map specifically for VTEP alone, but how that fit into ipcache map policy/identity...picture? any thought on this ? @joestringer

Reply:
At N=2 it's hard to say which approach is optimal, I wouldn't be surprised if a quick loop like this is less cycles than doing a hashmap lookup. But yeah I agree that a dedicated hashmap is a lot more flexible and doesn't need to be very complicated.

Reply:
@vincentmli originally we were concerned about putting all of this information into the ipcache value, as you could have thousands of ipcache map entries and we would be increasing the total memory usage based on needing just a small N (<10) number of entries about the tunnel content.

If we wanted to do this, I would suggest that the ipcache piece stays the same as now - basically identifying the VTEP IP to be used. Then in this conditional case, if there is a VTEP then use the VTEP IP to look up a separate map specifically with the other VTEP details like MAC addresses. This should keep the memory usage down, while also avoiding to execute these additional instructions when VTEP is disabled.

Reply:
> At N=2 it's hard to say which approach is optimal, I wouldn't be surprised if a quick loop like this is less cycles than doing a hashmap lookup.

I agree it probably won't be optimal for performance, but way less complexity issues and can scale to pretty much an arbitrary number of VTEPs.

> If we wanted to do this, I would suggest that the ipcache piece stays the same as now - basically identifying the VTEP IP to be used. Then in this conditional case, if there is a VTEP then use the VTEP IP to look up a separate map specifically with the other VTEP details like MAC addresses. This should keep the memory usage down, while also avoiding to execute these additional instructions when VTEP is disabled.

If there's a new map for VTEPs specifically, why is the ipcache entry needed? AFAIU, the ipcache entry is what is causing all the troubles with policies. If we simply have a dedicated map for VTEPs, it will work the same as the egress gateway or the SRv6 egress encapsulation.

Reply:
I opened an issue https://github.com/cilium/cilium/issues/18730 to track the discussion and for follow-up PR implementation, the PR got so many conversation and I could not easily find it :)

Reply:
Fair enough, my initial inclination to keep the entry in ipcache is mainly to avoid the extra VTEP map lookup based on a variable check from the ipcache entry (which we already look up anyway). We could alternatively just drop that piece and have a separate map and only enable that code based on enabling VTEP in the first place, so then you can compile the VTEP code out in the common case.

Reply:
if use VTEP specific map and remove VTEP from ipcache, I think we could still run into `toCIDR` policy issue with vtep CIDR https://github.com/cilium/cilium/issues/17694#issuecomment-1002195796

Reply:
We shouldn't run into that issue because VTEP handling and policy handling will then be completely separate (impact separate maps with no overlapping).

Reply:
Agree, the policy issue is introduced because of the VTEP handling code injecting the ipcache entry with identity 2. This is the piece that's unaware of other changes going on elsewhere in policy calculation which may set the identity differently.

Reply:
ok, I still have some doubt on `toCIDR`, but it sounds good to me, I am going to send a follow up PR for pod ARP proxying next since that would complete the beta feature of VTEP integration, then I will start to implement the VTEP BPF hash map and hopefully moving this beta feature to stable feature.